src/cmd_bundle_add.c

```diff
@@ -0,0 +1,176 @@
+	{ 0, 0, 0, 0 }
+};
+
+static bool parse_options(int argc, char **argv, struct list **opts)
+{
+	int opt;
+	bool list = false;
+
+	while ((opt = getopt_long(argc, argv, "hxu:c:v:P:p:F:lS:", prog_opts, NULL)) != -1) {
+		command_option_t *option = NULL;
+		int port = -1;
+		bool bool_true = true;
+
+		switch (opt) {
+		case '?':
+		case 'h':
+			print_help(argv[0]);
+			exit(EXIT_SUCCESS);
+		case 'u':
+			if (!optarg) {
+				printf("error: invalid --url argument\n\n");
+				return false;
+			}
+			option = construct_command_option("url", TYPE_STRING, optarg);
+			*opts = list_append_data(*opts, option);
+			break;
+		case 'c':
+			if (!optarg) {
+				printf("Invalid --contenturl argument\n\n");
+				return false;
+			}
+			option = construct_command_option("contenturl", TYPE_STRING, optarg);
+			*opts = list_append_data(*opts, option);
+			break;
+		case 'v':
+			if (!optarg) {
+				printf("Invalid --versionurl argument\n\n");
+				return false;
+			}
+			option = construct_command_option("versionurl", TYPE_STRING, optarg);
+			*opts = list_append_data(*opts, option);
+			break;
+		case 'p': /* default empty path_prefix verifies the running OS */
+			if (!optarg) {
+				printf("Invalid --path argument\n\n");
+				return false;
+			}
+			option = construct_command_option("path", TYPE_STRING, optarg);
+			*opts = list_append_data(*opts, option);
+			break;
+		case 'P':
+			if (sscanf(optarg, "%i", &port) != 1) {
+				printf("Invalid --port argument\n\n");
+				return false;
+			}
+			option = construct_command_option("port", TYPE_INT, &port);
+			*opts = list_append_data(*opts, option);
+			break;
+		case 'F':
+			if (!optarg || !is_format_correct(optarg)) {
+				printf("Invalid --format argument\n\n");
+				return false;
+			}
+			option = construct_command_option("format", TYPE_STRING, optarg);
+			*opts = list_append_data(*opts, option);
+			break;
+		case 'S':
+			if (!optarg || !is_statedir_correct(optarg)) {
+				printf("Invalid --statedir argument\n\n");
+				return false;
+			}
+			option = construct_command_option("statedir", TYPE_STRING, optarg);
+			*opts = list_append_data(*opts, option);
+			break;
+		case 'l':
+			list = true;
+			option = construct_command_option("list", TYPE_BOOL, &list);
+			*opts = list_append_data(*opts, option);
+			break;
+		case 'x':
+			option = construct_command_option("force", TYPE_BOOL, &bool_true);
+			*opts = list_append_data(*opts, option);
+			break;
+		default:
+			printf("error: unrecognized option\n\n");
+			return false;
+		}
+	}
+
+	if (!list) {
+		if (argc <= optind) {
+			printf("error: missing bundle(s) to be installed\n\n");
+			return false;
+		}
+	}
+
+	return true;
+}
+
+int bundle_add_main(int argc, char **argv)
+{
+	struct list *opts = NULL;
+	int ret = -1;
+
+	if (!parse_options(argc, argv, &opts)) {
+		print_help(argv[0]);
+		goto finish;
+	}
+
+        ret = dbus_client_call_method("BundleAdd", opts, DBUS_CMD_MULTIPLE_ARGS, (argv + optind));
```

Comment:
Actually the `DBUS_CMD_MULTIPLE_ARGS` option makes `dbus_client_call_method()` interpret `(argv + optind)` as a slice of the string vector `argv` from the `optind` position to its end. So we can add multiple bundles at once.

But I agree the code is such it may be not obvious from first sight. Added a comment to clarify what's going on here.